File: problem_015/problem_015_others3.pl
#require "memoize.rb"

sub find_routes {
  my ($x, $y) = @_;
  my $routes = 0;
  if ($y > 0) {
    $routes += find_routes($x, $y - 1);
  }
  if ($x > 0) {
    $routes += find_routes($x - 1, $y);
  }
  if ($x == 0 && $y == 0) {
    return 1;
  }
  return $routes;
}
#memoize :find_routes
#t = Time.now
#p find_routes(20, 20)
#t = Time.now - t
#p "elapsed time: %.2f" % t 
#print find_routes(20,20),"\n";

my $n=$ARGV[0];
print find_routes($n,$n),"\n";

File: problem_015/problem_015_others3.rb
#require "memoize.rb"
require "/home/vhv/.rvm/gems/ruby-1.9.2-p290/gems/i18n-0.5.0/lib/i18n/backend/memoize.rb"

def find_routes(x, y)
routes = 0
if y > 0
routes+=find_routes(x, y - 1)
end
if x > 0
routes+=find_routes(x - 1, y)
end
if x == 0 && y == 0
routes+=1
end
routes
end
memoize :find_routes #? not work!


t = Time.now
p find_routes(20, 20)
t = Time.now - t
p "elapsed time: %.2f" % t 
